### PR TITLE
Rename concatenate_tests.cu to .cpp

### PR DESCRIPTION
**Pull request by davidwendt**
Found this file can be changed to .cpp by modifying the `TypedColumnTest` to build its test data in the CPU and then cudaMemcpy it to the GPU. The remain ~1000 lines did not require any updates. Only a few `lists_column_wrapper` ctors needed to be corrected as similar to PR #8527.

### CODE REVIEW

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/copying/concatenate_tests.cpp [hunk 1]
```diff
@@ -59,8 +57,20 @@ struct TypedColumnTest : public cudf::test::BaseFixture {
   {
     auto typed_data = static_cast<char*>(data.data());
     auto typed_mask = static_cast<char*>(mask.data());
-    thrust::sequence(thrust::device, typed_data, typed_data + data_size());
-    thrust::sequence(thrust::device, typed_mask, typed_mask + mask_size());
+    std::vector<char> h_data(data_size());
+    thrust::sequence(h_data.begin(), h_data.end());
+    std::vector<char> h_mask(mask_size());
+    thrust::sequence(h_mask.begin(), h_mask.end());
```
Comment (harrism) [inline cpp/tests/copying/concatenate_tests.cpp:63]: I like to prefer `std` over `thrust` when it does the same thing.

```suggestion
    std::vector<char> h_data(data_size());
    std::iota(h_data.begin(), h_data.end());
    std::vector<char> h_mask(mask_size());
    std::iota(h_mask.begin(), h_mask.end());
```

File: cpp/tests/copying/concatenate_tests.cpp [hunk 2]
```diff
@@ -59,8 +57,20 @@ struct TypedColumnTest : public cudf::test::BaseFixture {
...
+    CUDA_TRY(cudaMemcpyAsync(typed_data,
```
Comment (harrism) [inline cpp/tests/copying/concatenate_tests.cpp:64]: I think you should either synchronize the stream right after these copies, or make them non-async `cudaMemcpy`.

File: cpp/tests/copying/concatenate_tests.cpp [hunk 3]
```diff
@@ -703,7 +713,7 @@ TEST_F(ListsColumnTest, ConcatenateEmptyLists)
   }
 
   {
-    cudf::test::lists_column_wrapper<int> a{{LCW{}}};
+    cudf::test::lists_column_wrapper<int> a{LCW{}};
```
Comment (harrism) [inline cpp/tests/copying/concatenate_tests.cpp:716]: I thought all these examples had all the nested brackets for a reason (nested lists?).

**Review by harrism — CHANGES_REQUESTED** [review]

File: cpp/tests/copying/concatenate_tests.cpp (hunk 3 above)
Comment (ttnghia) [reply cpp/tests/copying/concatenate_tests.cpp:716]: As I noticed, if you write `{LCW{}}` then the compiler optimizes it out to just `LCW{}`. If you want to write nested list, you have to explicitly write `LCW{LCW{}}` (add the class name).

Comment (davidwendt) [reply cpp/tests/copying/concatenate_tests.cpp:716]: The change was required to get past a runtime exception in `lists_column_wrapper`.  The extra `{}` were causing unexpected nested/depth levels. I worked with @nvdbaranec to work out the proper syntax for these.

Comment (ttnghia) [reply cpp/tests/copying/concatenate_tests.cpp:716]: Oh yeah, I forgot that there is also inconsistency between the compilers' behavior!

**Review by nvdbaranec — APPROVED** [review]

File: cpp/tests/copying/concatenate_tests.cpp [hunk 4]
```diff
@@ -51,26 +51,21 @@ struct TypedColumnTest : public cudf::test::BaseFixture {
...
-                             h_mask.data(),
-                             mask_size(),
-                             cudaMemcpyHostToDevice,
-                             rmm::cuda_stream_default.value()));
+    std::iota(h_mask.begin(), h_mask.end(), char{0});
+    CUDA_TRY(cudaMemcpyAsync(
+      typed_data, h_data.data(), data_size(), cudaMemcpyHostToDevice, stream.value()));
+    CUDA_TRY(cudaMemcpyAsync(
+      typed_mask, h_mask.data(), mask_size(), cudaMemcpyHostToDevice, stream.value()));
+    stream.synchronize();
```
Comment (harrism) [inline cpp/tests/copying/concatenate_tests.cpp:68]: You could have stuck with the default stream and just synced it, but this is more general. :+1:

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**